### ISSUE TITLE
关于GSConv模块源码中卷积核大小的问题

### ISSUE THREAD

**Issue by Squ1rrels** [AlanLi1997/slim-neck-by-gsconv#21]
您好，我在阅读源码时发现GSConv模块中的DWConv采用的时卷积核大小为5的卷积核，请问这里是特意采用k=5而不使用k=3的吗，我在yolov5上使用voc数据集测试过k=3和k=5的情况，发现map几乎是相同的，并且k=5时因为卷积核大小增大还会使参数量也变大，这里采用k=3会不会是更好的选择呢
```
class GSConv(nn.Module):
    # GSConv https://github.com/AlanLi1997/slim-neck-by-gsconv
    def __init__(self, c1, c2, k=1, s=1, g=1, act=True):
        super().__init__()
        c_ = c2 // 2
        self.cv1 = Conv(c1, c_, k, s, None, g, act)
        self.cv2 = Conv(c_, c_, 5, 1, None, c_, act)

    def forward(self, x):
        x1 = self.cv1(x)
        x2 = torch.cat((x1, self.cv2(x1)), 1)
        # shuffle
        y = x2.reshape(x2.shape[0], 2, x2.shape[1] // 2, x2.shape[2], x2.shape[3])
        y = y.permute(0, 2, 1, 3, 4)
        return y.reshape(y.shape[0], -1, y.shape[3], y.shape[4])
```

**Comment by AlanLi1997** [comment]
> 您好，我在阅读源码时发现GSConv模块中的DWConv采用的时卷积核大小为5的卷积核，请问这里是特意采用k=5而不使用k=3的吗，我在yolov5上使用voc数据集测试过k=3和k=5的情况，发现map几乎是相同的，并且k=5时因为卷积核大小增大还会使参数量也变大，这里采用k=3会不会是更好的选择呢
> 
> ```
> class GSConv(nn.Module):
>     # GSConv https://github.com/AlanLi1997/slim-neck-by-gsconv
>     def __init__(self, c1, c2, k=1, s=1, g=1, act=True):
>         super().__init__()
>         c_ = c2 // 2
>         self.cv1 = Conv(c1, c_, k, s, None, g, act)
>         self.cv2 = Conv(c_, c_, 5, 1, None, c_, act)
> 
>     def forward(self, x):
>         x1 = self.cv1(x)
>         x2 = torch.cat((x1, self.cv2(x1)), 1)
>         # shuffle
>         y = x2.reshape(x2.shape[0], 2, x2.shape[1] // 2, x2.shape[2], x2.shape[3])
>         y = y.permute(0, 2, 1, 3, 4)
>         return y.reshape(y.shape[0], -1, y.shape[3], y.shape[4])
> ```

你好，感谢你注意到一个论文中没被提及的很重要的细节。k=5是特意使用的，原因是：在表征学习中，我们期望允许模型拥有更丰富的特征多样性，以此来提升模型的表征能力。具体地说，K=5可以产生和K=3不同的纹理特征从而拓展特征多样性，并且深度卷积的K=5带来的参数量的增长是微弱的；另一个角度，K=5的深度卷积为CNNs模型带来了更大的感受野，对检测任务是更有利的。事实上，如果应用场景中对延时要求不高，你甚至可以将K提升至17(在我们的测试实验中，K=17时GSConv-Yolov5性能达到饱和)

**Comment by Squ1rrels** [comment]
> > 您好，我在阅读源码时发现GSConv模块中的DWConv采用的时卷积核大小为5的卷积核，请问这里是特意采用k=5而不使用k=3的吗，我在yolov5上使用voc数据集测试过k=3和k=5的情况，发现map几乎是相同的，并且k=5时因为卷积核大小增大还会使参数量也变大，这里采用k=3会不会是更好的选择呢
> > ```
> > class GSConv(nn.Module):
> >     # GSConv https://github.com/AlanLi1997/slim-neck-by-gsconv
> >     def __init__(self, c1, c2, k=1, s=1, g=1, act=True):
> >         super().__init__()
> >         c_ = c2 // 2
> >         self.cv1 = Conv(c1, c_, k, s, None, g, act)
> >         self.cv2 = Conv(c_, c_, 5, 1, None, c_, act)
> > 
> >     def forward(self, x):
> >         x1 = self.cv1(x)
> >         x2 = torch.cat((x1, self.cv2(x1)), 1)
> >         # shuffle
> >         y = x2.reshape(x2.shape[0], 2, x2.shape[1] // 2, x2.shape[2], x2.shape[3])
> >         y = y.permute(0, 2, 1, 3, 4)
> >         return y.reshape(y.shape[0], -1, y.shape[3], y.shape[4])
> > ```
> 
> 你好，感谢你注意到一个论文中没被提及的很重要的细节。k=5是特意使用的，原因是：在表征学习中，我们期望允许模型拥有更丰富的特征多样性，以此来提升模型的表征能力。具体地说，K=5可以产生和K=3不同的纹理特征从而拓展特征多样性，并且深度卷积的K=5带来的参数量的增长是微弱的；另一个角度，K=5的深度卷积为CNNs模型带来了更大的感受野，对检测任务是更有利的。事实上，如果应用场景中对延时要求不高，你甚至可以将K提升至17(在我们的测试实验中，K=17时GSConv-Yolov5性能达到饱和)

您好，首先非常感谢您的详细解答，我之所以会认为在GSConv模块中的DWConv采用k=3的卷积核或许是更好的选择，其实是源自Ghostnet论文中的这段原话“This is because that kernels of size 1 × 1 cannot introduce spatial information on feature maps, while larger kernels such as d = 5 or d = 7 lead to overfitting and more computations. Therefore, we adopt d = 3 in the following experiments for effectiveness and efficiency.”Ghostnet的作者认为d = 5或d = 7会导致过度拟合和更多计算，并且其在论文中的实验也证明了d = 3时网络性能最佳，因为你们提出的GSConv卷积是通过融合Ghost模块与Shuffle改进而来，GSConv模块中DW卷积卷积核大小也对应Ghostnet论文中所提到的d的大小，所以基于此我才会觉得Ghostnet结论或许可以泛化到GSConv中，不过或许论文中的“d = 3时网络性能最佳“观点仅限于他们所使用的数据集。

**Comment by AlanLi1997** [comment]
你好，我们必须注意到，DWConv中k的取值对不同规模的CNN模型的性能的影响是不同的。在轻量级CNN模型中使用k=5*5可直接带来明显的精度增益而只消耗较少的计算成本(性价比更高)，但随着CNN模型的增大，k=3*3和k=5*5对精度的贡献的差距会越来越小直到无差别，这时k=3*3便成为了更好的选择。

**Comment by AlanLi1997** [comment]
GSConv在形式上似乎是GhostNet+ShuffleNet，但其实这是一个有趣的巧合。我们一直在思考如何让低成本的DWConv获得和标准卷积一样好的效果，如何将通道密集卷积(标准卷积)的隐含知识**简单而直接**地传递给通道稀疏卷积——用 **通道混洗**融合**通道密集+通道稀疏** 是我们当时能想到的最优解。当然，这也直接表现为GhostNet+ShuffleNet🍺

**Comment by Squ1rrels** [comment]
感谢您的详细解答，您的解答对我很有帮助🍺
